src/osd/PrimaryLogPG.cc

```diff
@@ -2570,10 +2555,9 @@ int PrimaryLogPG::do_manifest_flush(OpRequestRef op, ObjectContextRef obc, Flush
     unsigned flags = CEPH_OSD_FLAG_IGNORE_CACHE | CEPH_OSD_FLAG_IGNORE_OVERLAY |
 		     CEPH_OSD_FLAG_RWORDERED;
     tgt_length = chunk_data.length();
-    if (pg_pool_t::fingerprint_t fp_algo = pool.info.get_fingerprint_type();
-	iter->second.has_reference() &&
-	fp_algo != pg_pool_t::TYPE_FINGERPRINT_NONE) {
-      object_t fp_oid = [fp_algo, &chunk_data]() -> string {
+    if (is_dedup_chunk(iter->second)) {
```

Comment:
There is no dirty entries. But, don't we need to flush newly updated chunk? do_manifest_flush() iterates over the chunk_map, then generate new fingerprint value for chunk OID. See below case.

Set-chunk object A (offset: 2~3. content: aaa)
Write object A (offset: 2~3, content: bbb)
At here, aaa's reference is dropped and the chunks's state is clean.
Flush object A triggers generating bbb's fingerprint ID and flush bbb to low tier.

You mean that we should prevent re-fingerprinting the object even if the chunk has the same content ?

Reply:
Your recent changes ensure that after that write, the state isn't clean, it's removed from the map completely.  I think what you want to do here is rerun cdc algorithm on the extents *not in the map* and then dedup those.

Reply:
Actually, if you agree with what I outlined above, I think you should remove this code, add a TODO comment, and address it in a future PR.

Reply:
Hm, I don't understand what you say here. 

> 1. Object A (content: aaaaaaaaaa)
> 2. Set-chunk object A (offset: 2-3, content: aaa)

At this point, how can we flush object A to the low tier without flush() ?
We need a tier_flush explicitly, right?

Or you mean that set-chunk needs to triggers a flush and make the chunk clean eventually?

Reply:
We agreed previously that set_chunk does not imply that flush will eventually re-write the chunk if it becomes overwritten.  There were several reasons for it, see our conversation above.

Thus, there are two different pathways by which extents of an object might become chunked:
1) An external agent runs a CDC and explicitly sends set-chunk commands for each chunk it chooses to dedup.
2) The osd internally runs a CDC on the extents of the object that are not yet dedup'd and performs the dedup directly.

1) so far as I can tell has absolutely nothing to do with this pathway.  If a chunk is dedup'd and then overwritten, its entry in the chunk_map is removed and it as if it *had not been dedup'd in the first place* (again, see our discussion previously in this PR for reasons).

Here, we deal with 2).  The starting state here is that the OSD has selected an object to be dedup'd.  Some extents in the object may already be dedup'd.  Those extents are the entries in chunk_map.  They are either present in the object or not.  In either case, I think that flush is a noop as those extents match the dedup targets.  The remainder of the object (chunks not in chunk_map) is what we might choose to run a CDC on and dedup.

I don't think that this implementation makes sense in the context of the current design (unless I'm really misunderstanding).  I think for now that you should remove this code entirely and submit a future PR with a more suitable design.

Reply:
> Hm, I don't understand what you say here.
> 
> > 1. Object A (content: aaaaaaaaaa)
> > 2. Set-chunk object A (offset: 2-3, content: aaa)
> 
> At this point, how can we flush object A to the low tier without flush() ?

We don't need to flush A at all.  The agent which sent the set-chunk already wrote the data to the target oid, all we have to do is update chunk_map and (depending on flag) take a reference, all of which set-chunk already does.

After a write

write object A (offset: 2-3, content bbb)

offset 2-3 is now removed from the chunk_map and the object is as if we had never dedup'd that extent at all (see above for our conversation about why this is desirable).

> We need a tier_flush explicitly, right?
> The tier_flush calls do_manifest_flush, then it will find out the chunk which should be flushed.

I think "flush" for a chunked object should cause the osd to re-run a CDC on the object, writeout new dedup targets (or take references), and update chunk_map to fit *for the portions of the object not already in the chunk_map*.

> is_dedup_chunk is something like need_dedup() -- Is this chunk a target of the dedup chunk?
> 
> I think the below case is the scenario we discussed. Am I missing something?
> 
>     1. Object A (content: aaaaa)
> 
>     2. run CDC
> 
>     3. Set-chunk object A (offset: 2-3, content: aaa)
> 
>     4. Flush object A

I'm not sure what you mean by flush here.  If you mean write out aaa to a chunk in the dedup pool, you definitely cannot do that *after* calling set-chunk, it would have to be before.  I'd argue that that would be part of 2.

> 
>     5. Write object A (content: aabbbb)

This completely removes the offset 2-3 chunk_map entry we created above.

> 
>     6. run CDC

This step should be identical to 2.  We run the CDC (again) on the whole object since the chunk_map is empty and write out the bbbb object to the dedup pool.  This sequence could happen either on an external agent or locally on the osd.

> 
>     7. Set-chunk objct A (offset: 2-4, content: bbbb)
>        ...

Having *already* written out the object containing bbbb, we use set-chunk to update the chunk_map.

> 
> 
> Or you mean that set-chunk should to be changes as it triggers a flush and make the chunk clean eventually?

chunks can either be clean or they can be missing.  They cannot be dirty.  Any chunk that exists in the chunk map must reference an object that really does have the same contents as the rados object should have at that offset and length.

Reply:
Understood (hm.. have we ever discussed this?). Anyway, what I should do now is adding a comment that describes "TODO". Then, remove this code entirely via submitting a future PR. Right?

Reply:
I'm referring to the comment chain at https://github.com/ceph/ceph/pull/35899#discussion_r466161216

Yeah, it's fine to do this in a seperate PR, just remove the code that is no longer valid and leave a // TODO with a brief explanation.

Reply:
@myoungwon Does this make sense to you?  I'm not so much arguing against your approach as I am genuinely not sure how it's meant to work, perhaps I'm missing something?

Reply:
@athanatos Yeah, It makes sense to me. One concern is that it is more complicated from user perspective.

In your scenario 1,
An external agent runs a CDC and explicitly sends set-chunk commands for each chunk it chooses to dedup.

1. Object A (content: aaaaa)
2. run CDC and write out the chunk (offset: 2-3, content: aaa)
3. Set-chunk object A (offset: 2-3, content: aaa)

The user probably needs copy-from or something in 2 to write out the object A to the low tier ?

Reply:
First, the "external agent" here could be run locally on each osd (in the existing cache/tiering agent thread, for instance).  Second, there's really no reason an actual external agent couldn't simply write out the object using a normal rados write -- keep in mind it already read in the object in order to do the CDC.  I don't see an obvious reason why you'd need something like copy-from.  Cache/tiering requires something like copy-from because it has excessively complicated atomicity requirements.  The same is not true here -- the dedup operation doesn't really happen (doesn't begin to affect user requests) until the set-chunk, which is obviously atomic and can be guarded with a version check.

Also, "user" here is misleading.  We're talking about another service that will be written and run automatically elsewhere using librados, not a human operator.

I'm also still really not sure what alternative you were proposing.

Reply:
>"external agent" here could be run locally on each osd (in the existing cache/tiering agent thread, for instance).

Ah, Ok.

>  keep in mind it already read in the object in order to do the CDC. 

Right.

>Also, "user" here is misleading. We're talking about another service that will be written and run automatically elsewhere using librados, not a human operator.

Ok.

> I'm also still really not sure what alternative you were proposing. Assuming we agree on #35899 (comment), it would seem that this is the natural approach.

Agree. I

Reply:
See my comment above.